src/sagemaker_huggingface_inference_toolkit/handler_service.py

```diff
@@ -272,3 +287,22 @@ def validate_and_initialize_user_module(self):
                 self.postprocess = postprocess_fn
             if transform_fn is not None:
                 self.transform_fn = transform_fn
+
+    def run_handler_function(self, func, *argv):
+        """Helper to call the handler function which covers 2 cases:
+        1. the handle function takes context
+        2. the handle function does not take context
+        """
+        num_func_input = len(signature(func).parameters)
+        if num_func_input == len(argv):
+            # function does not take context
+            result = func(*argv)
+        elif num_func_input == len(argv) + 1:
+            # function takes context
+            argv_context = argv + (self.context,)
+            result = func(*argv_context)
+        else:
+            raise TypeError(
+                "{} takes {} arguments but {} were given.".format(func.__name__, num_func_input, len(argv))
+            )
+        return result
```

Comment:
Is that needed? What overhead does it add to an request/function call?. 

Why don't we just past `self.context` in the `handle` method in line https://github.com/aws/sagemaker-huggingface-inference-toolkit/blob/44e3decd8aab4a710ef5f1094c39818cf7ea0f28/src/sagemaker_huggingface_inference_toolkit/handler_service.py#L234

Reply:
Hi @philschmid . Thank you for review. Maybe I am not understanding your point correctly, but I don't think we can directly pass `self.context` in the `handle` method. If we do that, don't we risk breaking existing customers who are not using `context` as an input argument? With the above mentioned `run_handler_function`, we should be able to support both customers who do not want to use `context` as well as customers who want to use `context`. Please correct me if I misunderstood. Thanks!

Reply:
Why would it break it? you mean if they have a custom `inference.py` that defines a `transform_fn` method?

Reply:
Yes, we might affect those who define a custom `transform_fn` right?

Reply:
In `handler_service.py`, we are trying to import user defined functions from `inference.py`

```
            load_fn = getattr(user_module, "model_fn", None)
            preprocess_fn = getattr(user_module, "input_fn", None)
            predict_fn = getattr(user_module, "predict_fn", None)
            postprocess_fn = getattr(user_module, "output_fn", None)
            transform_fn = getattr(user_module, "transform_fn", None)
```

If existence, it will be used to overwrite `self.load`, `self.preprocess`, `self.predict`, `self.postprocess` and `self.transform_fn`

```
            if load_fn is not None:
                self.load = load_fn
            if preprocess_fn is not None:
                self.preprocess = preprocess_fn
            if predict_fn is not None:
                self.predict = predict_fn
            if postprocess_fn is not None:
                self.postprocess = postprocess_fn
            if transform_fn is not None:
                self.transform_fn = transform_fn
```

In this PR, we introduce additional parameter `context` for default handlers. However, for existing user scripts, they don't have this parameter when they implement customized handler function. We have to be careful when we call these functions. That's why @sachanub is adding a helper function to determine when to call functions with the new parameter.

Reply:
Is that something we could deprecate and remove with a v3 of the toolkit? Feels like very error prone or a big overhead to parse and add args like this for every incoming request. 

What do you think of adding a check to see if there is a `inference.py` provided and if not we are using `self-transform` directly? Most customer deploy models using the "zero-code" deployment feature, where you provide a `MODEL_ID` and `TASK` and don't need and `inference.py` script.